component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -0,0 +1,272 @@
+                } catch (NumberFormatException e) {
+                    validityPeriod = TokenExchangeConstants.DEFAULT_IAT_VALIDITY_PERIOD;
+                    log.warn("Invalid value: " + validityPeriodProp + " is set for IAT validity period. Using " +
+                            "default value: " + validityPeriod + " minutes.");
+                }
+            } else {
+                validityPeriod = TokenExchangeConstants.DEFAULT_IAT_VALIDITY_PERIOD;
+                log.warn("Empty value is set for IAT validity period. Using default value: " + validityPeriod
+                        + " minutes.");
+            }
+        }
+
+        String registeredClaims = IdentityUtil.getProperty(TokenExchangeConstants.REGISTERED_CLAIMS);
+        if (StringUtils.isNotBlank(registeredClaims)) {
+            registeredClaimNames = registeredClaims.split("\\s*,\\s*");
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Validate IAT is set to: " + validateIAT + " for JWT grant.");
+            if (validateIAT) {
+                log.debug("IAT validity period is set to: " + validityPeriod + " minutes for JWT grant.");
+            }
+        }
+    }
+
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+        SignedJWT signedJWT;
+        IdentityProvider identityProvider;
+        String tokenEndPointAlias;
+        JWTClaimsSet claimsSet = null;
+        String requested_audience = null;
+        boolean audienceFound;
+
+        RequestParameter[] params = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getRequestParameters();
+        Map<String, String[]> requestParams = Arrays.stream(params).collect(Collectors.toMap(RequestParameter::getKey,
+                RequestParameter::getValue));
+        if (requestParams.get(TokenExchangeConstants.REQUESTED_TOKEN_TYPE) != null) {
+            requested_token_type = requestParams.get(TokenExchangeConstants.REQUESTED_TOKEN_TYPE)[0];
+        }
+        if(requestParams.get(TokenExchangeConstants.AUDIENCE) != null) {
+            requested_audience = requestParams.get(TokenExchangeConstants.AUDIENCE)[0];
+        }
+        String tenantDomain = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain();
+        if (StringUtils.isEmpty(tenantDomain)) {
+            tenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+        }
+        if (TokenExchangeConstants.JWT_TOKEN_TYPE.equals(requested_token_type)) {
+            signedJWT = TokenExchangeUtils.getSignedJWT(requestParams.get(TokenExchangeConstants.SUBJECT_TOKEN)[0]);
+            if (signedJWT == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "No Valid subject token was found for "
+                        + TokenExchangeConstants.TOKEN_EXCHANGE_GRANT_TYPE);
+            } else {
+                claimsSet = TokenExchangeUtils.getClaimSet(signedJWT);
```

Comment:
Any reasons for moving getClaimSet to a util method? If is is not used in multiple places`signedJWT.getJWTClaimsSet` , we can have it here and get rid of util

Reply:
The initial implementation of token exchange is planned to be compatible with IS 5.11. We decided to use the methods inside TokenExchangeUtils class (will result in code duplication for now) and this will be removed in the future when upgrading to 5.12 compatible dependencies.

Since, the same logic is there in the JWT bearer grant implementation, it would be better to add this to a common util and use the same if it's not there already